src/wasm/wasm-emscripten.cpp

```diff
@@ -587,45 +600,109 @@ void AsmConstWalker::noteNonLinear(Expression* curr) {
 
 void AsmConstWalker::visitLocalSet(LocalSet* curr) { sets[curr->index] = curr; }
 
+Const* AsmConstWalker::resolveConstAddr(Expression* arg, const Name& target) {
+  while (!arg->dynCast<Const>()) {
+    if (auto* get = arg->dynCast<LocalGet>()) {
+      // The argument may be a local.get, in which case, the last set in this
+      // basic block has the value.
+      auto* set = sets[get->index];
+      if (set) {
+        assert(set->index == get->index);
+        arg = set->value;
+      }
+    } else if (auto* value = arg->dynCast<Binary>()) {
+      // In the dynamic linking case the address of the string constant
+      // is the result of adding its offset to __memory_base.
+      // In this case are only looking for the offset with the data segment so
+      // the RHS of the addition is just what we want.
+      assert(value->op == AddInt32);
```

Comment:
We're assuming things about the nature of the generated code. Is there more we can assert about the LHS here?

Reply:
This is the way the code was originally written. I think it's better to have more assertions, but I don't think it's in scope for this PR.

Reply:
Ah right, method extraction, carry on